src/storage/engine/mod.rs

```diff
@@ -81,96 +81,237 @@ pub trait Snapshot: Send {
+
+pub struct Cursor<'a> {
+    iter: Box<Iterator + 'a>,
+    scan_mode: ScanMode,
+    // the data cursor can be seen will be
+    min_key: Option<Vec<u8>>,
+    max_key: Option<Vec<u8>>,
+}
+
+impl<'a> Cursor<'a> {
+    pub fn new<T: Iterator + 'a>(iter: T, mode: ScanMode) -> Cursor<'a> {
+        Cursor {
+            iter: Box::new(iter),
+            scan_mode: mode,
+            min_key: None,
+            max_key: None,
+        }
+    }
+
+    pub fn seek(&mut self, key: &Key) -> Result<bool> {
+        if self.max_key.as_ref().map_or(false, |k| k <= key.encoded()) {
+            try!(self.iter.validate_key(key));
+            return Ok(false);
+        }
+
+        if self.scan_mode == ScanMode::Forward && self.valid() && self.iter.key() >= key.encoded() {
+            return Ok(true);
+        }
+
+        if !try!(self.iter.seek(key)) {
+            self.max_key = Some(key.encoded().to_owned());
+            return Ok(false);
+        }
+        Ok(true)
+    }
 
     /// Seek the specified key.
     ///
     /// This method assume the current position of cursor is
     /// around `key`, otherwise you should use `seek` instead.
-    fn near_seek(&mut self, key: &Key) -> Result<bool> {
-        if !self.valid() {
+    pub fn near_seek(&mut self, key: &Key) -> Result<bool> {
+        if !self.iter.valid() {
             return self.seek(key);
         }
-        if self.key() == &**key.encoded() {
+        let ord = self.iter.key().cmp(key.encoded());
+        if ord == Ordering::Equal ||
+           (self.scan_mode == ScanMode::Forward && ord == Ordering::Greater) {
             return Ok(true);
         }
-        let (nav, ord): (fn(&mut _) -> bool, Ordering) = if self.key() > &**key.encoded() {
-            (Cursor::prev, Ordering::Greater)
-        } else {
-            (Cursor::next, Ordering::Less)
-        };
-        let mut cnt = 0;
-        while self.key().cmp(key.encoded()) == ord && nav(self) {
-            cnt += 1;
-            if cnt >= SEEK_BOUND {
-                CURSOR_OVER_SEEK_BOUND_COUNTER.inc();
-
-                return self.seek(key);
-            }
+        if self.max_key.as_ref().map_or(false, |k| k <= key.encoded()) {
+            try!(self.iter.validate_key(key));
+            return Ok(false);
         }
         if ord == Ordering::Greater {
-            if self.valid() {
-                if self.key() < &**key.encoded() {
-                    self.next();
+            near_loop!(self.iter.prev() && self.iter.key() > key.encoded(),
+                       self.seek(key));
+            if self.iter.valid() {
+                if self.iter.key() < key.encoded() {
+                    self.iter.next();
                 }
+            } else if self.iter.seek_to_first() {
+                return Ok(true);
             } else {
-                self.seek_to_first();
+                self.max_key = Some(key.encoded().to_owned());
+                if self.min_key.as_ref().map_or(true, |k| k > key.encoded()) {
```

Comment:
Update `min_key` when new min_key is greater then current min_key, not the opposite.